### PR TITLE
Add Relay for Mainnet -> xDai

### LINKED ISSUE #1606: Build out seamless xDai integration, perp.exchange style
This will enable users to enjoy the scalability of xDai without having to switch networks in their wallet provider.

### PR DESCRIPTION
**Pull request by hexyls**
closes #1606

Current iteration of the proxy factory that accepts signatures instead of auto approved hashes: [0x7b9756f8A7f4208fE42FE8DE8a8CC5aA9A03f356](https://blockscout.com/poa/xdai/address/0x7b9756f8A7f4208fE42FE8DE8a8CC5aA9A03f356/contracts)

### CODE REVIEW

File: app/src/hooks/connectedWeb3.tsx
Comment (Mi-Lan) [inline app/src/hooks/connectedWeb3.tsx:49]: Also with this pr there will be no option to see mainnet markets? I see that we have toogleRelay but its not used anywhere..

**Review by Mi-Lan — CHANGES_REQUESTED** [review]
Great job Hexyl nicely done! Love this feature. Do we have plans on how to send that initial transaction to xDai and fund the cpk? Left some more comments below as well...

File: app/src/hooks/connectedWeb3.tsx
Comment (hexyls) [reply app/src/hooks/connectedWeb3.tsx:49]: Yep there is no toggle UI yet we need to add this later 👍

File: app/src/hooks/connectedWeb3.tsx
Comment (hexyls) [reply app/src/hooks/connectedWeb3.tsx:112]: Yep this will all be sorted when we add the new UI 👍

**Comment by hexyls** [comment]
@Mi-Lan yep I have modified some of the original bridge code to work with the relay, so when @KadenZipfel sets up https://github.com/protofire/omen-exchange/issues/1680 he should be able to coordinate deposits.

File: app/src/services/biconomy.ts
Comment (KadenZipfel) [inline app/src/services/biconomy.ts:36]: Does having this visible leave us vulnerable to a DoS attack?

**Review by KadenZipfel — APPROVED** [review]
Pure wizardry 🧙‍♂️ awesome work! You can set the base to this branch: https://github.com/protofire/omen-exchange/pull/1698 where I'll integrate it with the front-end updates

File: app/src/services/biconomy.ts
Comment (hexyls) [reply app/src/services/biconomy.ts:36]: Yeah definitely, it looks like the only preventative are these daily limits that can be set in the Biconomy dashboard:

Screen Shot 2021-03-03 at 10 31 52 am

it looks like we can whitelist the proxy factory contract, but even then someone could spam the proxy factory... 🤔

Comment (hexyls) [reply app/src/services/biconomy.ts:36]: So I'm unsure if the user limit is per address or per IP address, can't find much info about it. Even per IP address an attacker can use rotating proxies I guess.

hmmmm soooo yeah potential attack - spam until dapp limit is depleted and no one can use the app. Good catch @KadenZipfel 

Will think about how we can obfuscate the api key.. might have deploy a tiny api between omen and biconomy.

Comment (zezimaSudo) [reply app/src/services/biconomy.ts:36]: why can't we put the API key in a process.env file and then add it to .gitignore?

Comment (hexyls) [reply app/src/services/biconomy.ts:36]: good idea @gem1no but this would still expose the api key to an attacker, they could snoop it from the network tab in chrome tools when requests are made for example.

Comment (tomarsachin2271) [reply app/src/services/biconomy.ts:36]: Per user API limit is on user address not IP address and apart from Limit module, you can add the domain restrictions as well..
[https://docs.biconomy.io/guides/biconomy-dashboard#domain-restrictions](https://docs.biconomy.io/guides/biconomy-dashboard#domain-restrictions)

So only requests from your domain will be entertained for your API Key but still there are chances of someone impersonating the domain origin of the request.

As @hexyls  suggested, best way would be to deploy a tiny api between omen and biconomy to completely obfuscate api key.

**Review by pimato — CHANGES_REQUESTED** [review]
It looks you took the bridge interface for xdai connection and applied it for ethereum mainnet:
Bildschirmfoto 2021-03-03 um 21 35 20

It should be DAI -> xDAI

**Comment by pimato** [comment]
I want to deposit 100 DAI: 
Bildschirmfoto 2021-03-03 um 21 36 45

after clicking transfer:
Bildschirmfoto 2021-03-03 um 21 36 49

I am getting a signing modal from metamask (I don`t think you can deposit dai to xdai via signing):
Bildschirmfoto 2021-03-03 um 21 37 02

**Comment by hexyls** [comment]
Yep @pimato ignore all UI stuff for now - my bad just looking for code review atm prob shouldn't have tagged you for review!

**Review by Mi-Lan — COMMENTED** [review]
Screenshot 2021-03-15 at 12 53 48
I think this margin should be 12px according to design

**Comment by Mi-Lan** [comment]
Screenshot 2021-03-15 at 12 59 23
Should we include icon's here ? Like in figma
Screenshot 2021-03-15 at 13 01 21

**Comment by Mi-Lan** [comment]
This padding should be removed...
Screenshot 2021-03-15 at 13 07 09

**Comment by Mi-Lan** [comment]
Should be 16px
Screenshot 2021-03-15 at 13 17 07

**Comment by Mi-Lan** [comment]
Seems that this text is different then in figma
Screenshot 2021-03-15 at 13 22 10
In figma:

> "To Deposit or Withdraw Dai to your Omen Account, you need to enable it first"

**Comment by Mi-Lan** [comment]
In figma these all margins are 20px except top which is 24px...
Screenshot 2021-03-15 at 13 25 12

**Comment by Mi-Lan** [comment]
When i click on enable then cancel the transaction this modal is stuck in infinite loading...I think it should close if user declines transaction...
Screenshot 2021-03-15 at 13 27 26

**Comment by Mi-Lan** [comment]
I would maybe expand the onClick to be on the wrapper of network text. 
Since it happened to me couple of times that i clicked on it and it closed but didn't switch network..

https://user-images.githubusercontent.com/33226956/111157017-ac76fd80-8596-11eb-8c86-9b82a7ee41cd.mov

**Comment by Mi-Lan** [comment]
Max button doesn't seem to work on deposit.

https://user-images.githubusercontent.com/33226956/111163367-58234c00-859d-11eb-9636-d3e2ff8652cb.mov

**Review by Mi-Lan — COMMENTED** [review]
I think you can deposit any amount on xDai there is only limit when you want to withdraw. So this should be only displayed for withdrawing.
Screenshot 2021-03-15 at 15 45 10

**Comment by pimato** [comment]
The Account button should in general show the native blockchain token balance. We have that for xDai but not for Ethereum Mainnet:

account-button

**Comment by pimato** [comment]
I am currently connected to Ethereum Mainnet, but don´t see the Omen Account component in the `Your Connection - Modal`: 

Bildschirmfoto 2021-03-15 um 16 26 25

That is probably a bug?

**Comment by pimato** [comment]
> When i click on enable then cancel the transaction this modal is stuck in infinite loading...I think it should close if user declines transaction...
> Screenshot 2021-03-15 at 13 27 26

Also, can we not overlap modals on top of each other?
Currently:
Bildschirmfoto 2021-03-15 um 16 43 51

We should hide the account modal until we are done with the transaction.

**Comment by Mi-Lan** [comment]
Also when i click on claim amount then cancel the transaction i get this error...I handled that with my bridge dunno why is it showing here?
Screenshot 2021-03-15 at 17 39 06

**Comment by hexyls** [comment]
Thanks for the great feedback @Mi-Lan @pimato 

The only thing I couldn't reproduce was the max button issue 🤔

**Comment by Mi-Lan** [comment]
> Thanks for the great feedback @Mi-Lan @pimato
> 
> The only thing I couldn't reproduce was the max button issue 🤔



> Thanks for the great feedback @Mi-Lan @pimato
> 
> The only thing I couldn't reproduce was the max button issue 🤔

Yeah and i really like how this is turning out! That feature to switch network on switch is smooth as hell! Hmm maybe my wording is bad max button does work but it isn't displaying amount properly...Will look into it myself! Maybe i can solve it

File: app/src/components/modal/modal_deposit_withdraw/index.tsx
Comment (Mi-Lan) [inline app/src/components/modal/modal_deposit_withdraw/index.tsx:198]: Small thing...i think that TextfieldcustomPlaceholder and bigNumber input can be self closing tag...

File: app/src/components/modal/modal_deposit_withdraw/index.tsx
Comment (Mi-Lan) [inline app/src/components/modal/modal_deposit_withdraw/index.tsx:201]: Also all these decimals could be imported from standard decimals Kaden created? Or is it not a good fit

**Comment by Mi-Lan** [comment]
@hexyls i took a stab at it but couldn't figure out how to solve the max button display issue...very weird doesn't make too much sense why its not working....

**Review by KadenZipfel — CHANGES_REQUESTED** [review]
Looking 🔥

**Comment by hexyls** [comment]
@Mi-Lan I finally figured out what is going on with that max balance issue, I made some sketchy changes to some formatting utilities earlier that were messing up the BigNumberInput component 😄 resolved now I think 

Haven't re-requested reviews yet because I realized we need a way to refresh balances in the header after buying/selling etc which I will work on tomorrow.

**Comment by Mi-Lan** [comment]
Max button works!!! Also one more thing! I just tested deposit feature and when the transaction finished and has 9 confirmations and i got back i didn't see my Omen account balance change...Only when i refreshed the page did it change.  For the bridge i implemented that when it does transaction and after it's confirmed it refetches balances...I think it will work in this case as well on deposit function...

**Comment by Mi-Lan** [comment]
Also when i spend something on some market balance in the top doesn't change only when i refresh the page it does...

https://user-images.githubusercontent.com/33226956/111515848-906c8b00-8753-11eb-89ef-06f45da6f34f.mov

File: app/src/components/common/layout/header/index.tsx
Comment (KadenZipfel) [inline app/src/components/common/layout/header/index.tsx:204]: Maybe better to just use `history.replace('/')` since it will redirect to /liquidity anyway and if we choose to change the routes in the future it won't cause any issues

**Review by KadenZipfel — CHANGES_REQUESTED** [review]
Nice work with the balances hook 🔥

**Comment by hexyls** [comment]
Alrighty, balances should update in the header after buys/sells etc now 🥳

Keep an eye out for major issues in the next review. We are trying to merge this for the next release, so if there are minor issues absolutely let me know and I will make seperate issues to work on after merge 👍

**Comment by pimato** [comment]
Testing:

General
- [ ] Redeem Dai on Ethereum Mainnet
- [ ] Deposit Dai into Omen

Scalar Market
- [ ] Create Market
- [ ] Buy outcome token
- [ ] Sell outcome token
- [ ] Deposit Liquidity
- [ ] Remove Liquidity

Categorical Market
- [ ] Create Market
- [ ] Buy outcome token
- [ ] Sell outcome token
- [ ] Deposit Liquidity
- [ ] Remove Liquidity

**Review by pimato — APPROVED** [review]
Testing:

General
- [x] Redeem Dai on Ethereum Mainnet
- [x] Deposit Dai into Omen

Scalar Market
- [x] Create Market
- [x] Buy outcome token
- [x] Sell outcome token
- [x] Deposit Liquidity
- [x] Remove Liquidity

Categorical Market
- [x] Create Market
- [x] Buy outcome token
- [x] Sell outcome token
- [x] Deposit Liquidity
- [x] Remove Liquidity

**Comment by Mi-Lan** [comment]
@hexyls i found that on this market(http://localhost:3000/#/0xfbca86af4c47bc91aacef2add882db4ce520bb56) i used for testing the scalar finalization i can't resolve condition...I gives out error 
Screenshot 2021-03-18 at 12 44 26

> here was an error resolving the condition with question id '0xbb90e7036bdde35da06e9aedc349dc7fba74947786a5c4dd6276c25427461323' o.signer.sendTransaction is not a function

**Comment by pimato** [comment]
> @hexyls i found that on this market(http://localhost:3000/#/0xfbca86af4c47bc91aacef2add882db4ce520bb56) i used for testing the scalar finalization i can't resolve condition...I gives out error
> Screenshot 2021-03-18 at 12 44 26
> 
> > here was an error resolving the condition with question id '0xbb90e7036bdde35da06e9aedc349dc7fba74947786a5c4dd6276c25427461323' o.signer.sendTransaction is not a function

can replicate it:
Bildschirmfoto 2021-03-18 um 18 18 07

**Review by pimato — CHANGES_REQUESTED** [review]
redeem issue

**Review by pimato — APPROVED** [review]
All works now!

**Review by Mi-Lan — APPROVED** [review]
GJ ! Works like charm!

https://user-images.githubusercontent.com/33226956/111769792-e81d0a80-88a9-11eb-8323-73df8c79f790.mov

**Review by KadenZipfel — APPROVED** [review]
🚀

**Pull request merged**
